src/useTextWidth.tsx

```diff
@@ -1,44 +1,57 @@
+
+export type useTextWidthTextOptions = {
+  text: string | string[];
+  font?: string;
+};
+
+export type useTextWidthRefOptions = {
+  ref: RefObject<Element>;
+};
 
 export interface useTextWidthType {
-  ({ text, font, ref }: { text?: string | string[]; font?: string; ref?: any }): number;
+  (options: useTextWidthTextOptions | useTextWidthRefOptions): number;
 }
 
-const useTextWidth: useTextWidthType = ({ text, ref, font = '16px Times' }) => {
-  const [width, setWidth] = useState(-1);
+const getContext = () => {
+  const fragment: DocumentFragment = document.createDocumentFragment();
+  const canvas: HTMLCanvasElement = document.createElement('canvas');
+  fragment.appendChild(canvas);
+  return canvas.getContext('2d') as CanvasRenderingContext2D;
+};
 
-  useEffect(() => {
-    const getContext = () => {
-      const fragment: DocumentFragment = document.createDocumentFragment();
-      const canvas: HTMLCanvasElement = document.createElement('canvas');
-      fragment.appendChild(canvas);
-      return canvas.getContext('2d');
-    };
+const getTextWidth = (currentText: string | string[], font: string) => {
+  const context = getContext();
+  context.font = font;
 
-    const getTextWidth = (currentText: string | string[]) => {
-      const context = getContext();
-      context.font = font;
+  if (Array.isArray(currentText)) {
+    return Math.max(...currentText.map((t) => context.measureText(t).width));
+  } else {
+    const metrics = context.measureText(currentText);
+    return metrics.width;
+  }
+};
 
-      if (Array.isArray(currentText)) {
-        return Math.max(...currentText.map((t) => context.measureText(t).width));
-      } else {
-        const metrics = context.measureText(currentText);
-        return metrics.width;
-      }
-    };
+const useTextWidth: useTextWidthType = (options) => {
+  const textOptions = useMemo(() => ('text' in options ? options : undefined), [options]);
+  const refOptions = useMemo(() => ('ref' in options ? options : undefined), [options]);
+
+  return useMemo(() => {
+    if (refOptions?.ref) {
+      if (!refOptions.ref.current || refOptions.ref.current.textContent === null) return NaN;
 
-    if (ref && ref.current) {
       const context = getContext();
-      const computedStyles = window.getComputedStyle(ref.current);
+      const computedStyles = window.getComputedStyle(refOptions.ref.current);
       context.font = computedStyles.font;
-      const metrics = context.measureText(ref.current.textContent);
-      setWidth(metrics.width);
-    } else if (text) {
-      setWidth(getTextWidth(text));
+      const metrics = context.measureText(refOptions.ref.current.textContent);
+
+      return metrics.width;
+    } else if (textOptions?.text) {
+      return getTextWidth(textOptions.text, textOptions.font ?? '16px times');
     }
-  }, [font, ref, text]);
 
-  return width;
+    throw new TypeError('useTextWidth - Either `ref` OR `text` must be defined');
```

Comment:
I think this might cause unwanted error messages in the console when the text/ref assigned but the actual value is still undefined or null. This is a common case during the application init phase. Also people should be aware of the hook argument needs to be provided since it's the most important thing here. Throwing exception here won't help for good IMO.

Reply:
Ref's should still have a `.current` even if they haven't been assigned, so they wouldn't encounter this unless there is some funky manipulation going on.

As far as Text, in strict mode projects they would be warned by the interface that `undefined` is not a valid value for `text`
That said we could loosen up that and return NaN like I have setup for when `ref.current` is null.